lib/ecto/adapters/sql.ex

```diff
@@ -245,6 +245,31 @@ defmodule Ecto.Adapters.SQL do
     end
   end
 
+  @doc """
+  TODO: adapter's specific details should be documented here?
+
+  ## Examples
+
+      iex> Ecto.Adapters.SQL.explain(:all, repo, Post)
+      "Seq Scan on posts p0  (cost=0.00..12.12 rows=1 width=443)"
+
+  """
+  @spec explain(Ecto.Repo.t, :all | :update_all | :delete_all, Ecto.Queryable.t, Keyword.t) :: String.t
+  def explain(repo, operation, queryable, opts \\ []) do
+    {prepared, params} = to_sql(operation, repo, queryable)
+    %{sql: sql} = adapter_meta = Ecto.Adapter.lookup_meta(repo)
+
+    {:ok, explain} =
+      transaction(adapter_meta, [], fn ->
```

Comment:
Do we need to start a transaction?

Reply:
I think so, to avoid side effects. Please see the topic "Important" here https://www.postgresql.org/docs/current/sql-explain.html

"Keep in mind that the statement is actually executed when the ANALYZE option is used. Although EXPLAIN will discard any output that a SELECT would return, other side effects of the statement will happen as usual."

Reply:
I see. Don't forget to rollback the transaction then at the end. :)

Reply:
I couldn't find a way to get the data executed between the savepoints (before the rollback) to return here (not without changing db_connection), because triggering a rollback will of course return a `{:error, reason}`, but turns out it's indeed calling a rollback even without calling it explicitly. See the tests https://github.com/elixir-ecto/ecto_sql/pull/231/commits/6c0e84473450cadb3f9bf9550b5fd0a6327c958c#diff-a233029a670aec75d70382c3f0330a47R143-R163 and the logs for delete and update respectively:

```
2020-05-27 18:28:23.268 EDT [88435] LOG:  statement: BEGIN
2020-05-27 18:28:23.279 EDT [88435] LOG:  statement: SAVEPOINT postgrex_query
2020-05-27 18:28:23.281 EDT [88435] LOG:  execute ecto_insert_posts: INSERT INTO "posts" ("public","title","uuid","inserted_at","updated_at") VALUES ($1,$2,$3,$4,$5) RETURNING "id"
2020-05-27 18:28:23.281 EDT [88435] DETAIL:  parameters: $1 = 't', $2 = 'hello', $3 = 'f2300309-3c7a-4f61-97b6-942291c3b896', $4 = '2020-05-27 22:28:23', $5 = '2020-05-27 22:28:23'
2020-05-27 18:28:23.281 EDT [88435] LOG:  statement: RELEASE SAVEPOINT postgrex_query
2020-05-27 18:28:23.284 EDT [88435] LOG:  statement: SAVEPOINT postgrex_query
2020-05-27 18:28:23.284 EDT [88435] LOG:  execute ecto_1476: SELECT p0."id", p0."counter", p0."title", p0."blob", p0."public", p0."cost", p0."visits", p0."wrapped_visits", p0."intensity", p0."bid", p0."uuid", p0."meta", p0."links", p0."intensities", p0."posted", p0."author_id", p0."inserted_at", p0."updated_at" FROM "posts" AS p0
2020-05-27 18:28:23.284 EDT [88435] LOG:  statement: RELEASE SAVEPOINT postgrex_query
2020-05-27 18:28:23.286 EDT [88435] LOG:  statement: SAVEPOINT postgrex_query
2020-05-27 18:28:23.286 EDT [88435] LOG:  execute <unnamed>: SHOW server_version
2020-05-27 18:28:23.286 EDT [88435] LOG:  statement: RELEASE SAVEPOINT postgrex_query
2020-05-27 18:28:23.286 EDT [88435] LOG:  statement: SAVEPOINT postgrex_savepoint
2020-05-27 18:28:23.287 EDT [88435] LOG:  execute <unnamed>: EXPLAIN ( ANALYZE FALSE, VERBOSE FALSE, COSTS TRUE, BUFFERS FALSE, TIMING FALSE, SUMMARY FALSE, SETTINGS FALSE ) DELETE FROM "posts" AS p0
2020-05-27 18:28:23.287 EDT [88435] LOG:  statement: RELEASE SAVEPOINT postgrex_savepoint
2020-05-27 18:28:23.287 EDT [88435] LOG:  statement: SAVEPOINT postgrex_query
2020-05-27 18:28:23.287 EDT [88435] LOG:  execute ecto_1476: SELECT p0."id", p0."counter", p0."title", p0."blob", p0."public", p0."cost", p0."visits", p0."wrapped_visits", p0."intensity", p0."bid", p0."uuid", p0."meta", p0."links", p0."intensities", p0."posted", p0."author_id", p0."inserted_at", p0."updated_at" FROM "posts" AS p0
2020-05-27 18:28:23.288 EDT [88435] LOG:  statement: RELEASE SAVEPOINT postgrex_query
2020-05-27 18:28:23.288 EDT [88435] LOG:  statement: ROLLBACK
```

```
2020-05-27 18:33:24.199 EDT [88846] LOG:  statement: BEGIN
2020-05-27 18:33:24.211 EDT [88846] LOG:  statement: SAVEPOINT postgrex_query
2020-05-27 18:33:24.213 EDT [88846] LOG:  execute ecto_insert_posts: INSERT INTO "posts" ("public","title","uuid","inserted_at","updated_at") VALUES ($1,$2,$3,$4,$5) RETURNING "id"
2020-05-27 18:33:24.213 EDT [88846] DETAIL:  parameters: $1 = 't', $2 = 'hello', $3 = 'ace2bd5b-a98b-4beb-9386-7176486822cc', $4 = '2020-05-27 22:33:24', $5 = '2020-05-27 22:33:24'
2020-05-27 18:33:24.213 EDT [88846] LOG:  statement: RELEASE SAVEPOINT postgrex_query
2020-05-27 18:33:24.216 EDT [88846] LOG:  statement: SAVEPOINT postgrex_query
2020-05-27 18:33:24.217 EDT [88846] LOG:  execute ecto_1096: SELECT p0."id", p0."counter", p0."title", p0."blob", p0."public", p0."cost", p0."visits", p0."wrapped_visits", p0."intensity", p0."bid", p0."uuid", p0."meta", p0."links", p0."intensities", p0."posted", p0."author_id", p0."inserted_at", p0."updated_at" FROM "posts" AS p0
2020-05-27 18:33:24.217 EDT [88846] LOG:  statement: RELEASE SAVEPOINT postgrex_query
2020-05-27 18:33:24.218 EDT [88846] LOG:  statement: SAVEPOINT postgrex_query
2020-05-27 18:33:24.218 EDT [88846] LOG:  execute <unnamed>: SHOW server_version
2020-05-27 18:33:24.218 EDT [88846] LOG:  statement: RELEASE SAVEPOINT postgrex_query
2020-05-27 18:33:24.219 EDT [88846] LOG:  statement: SAVEPOINT postgrex_savepoint
2020-05-27 18:33:24.219 EDT [88846] LOG:  execute <unnamed>: EXPLAIN ( ANALYZE FALSE, VERBOSE FALSE, COSTS TRUE, BUFFERS FALSE, TIMING FALSE, SUMMARY FALSE, SETTINGS FALSE ) UPDATE "posts" AS p0 SET "title" = 'new title'
2020-05-27 18:33:24.219 EDT [88846] LOG:  statement: RELEASE SAVEPOINT postgrex_savepoint
2020-05-27 18:33:24.219 EDT [88846] LOG:  statement: SAVEPOINT postgrex_query
2020-05-27 18:33:24.219 EDT [88846] LOG:  execute ecto_1096: SELECT p0."id", p0."counter", p0."title", p0."blob", p0."public", p0."cost", p0."visits", p0."wrapped_visits", p0."intensity", p0."bid", p0."uuid", p0."meta", p0."links", p0."intensities", p0."posted", p0."author_id", p0."inserted_at", p0."updated_at" FROM "posts" AS p0
2020-05-27 18:33:24.219 EDT [88846] LOG:  statement: RELEASE SAVEPOINT postgrex_query
2020-05-27 18:33:24.220 EDT [88846] LOG:  statement: ROLLBACK
```

Reply:
Hey @josevalim my mistake, it was calling ROLLBACK because of the test transaction. I fixed it by using `Ecto.Multi` so I don't need to to play with internal functions avoiding the risk of missing edge cases already solved by multi. The tricky here is to be able to obtain the `EXPLAIN` result even after triggering a rollback. Wdyt?

Reply:
I believe you can pass a value when calling rollback. :) Such as `Repo.rollback({:explain, result})`. But using multi works too.

Reply:
That was my first attempt, to `catch` that result that's throw'ed by rollback but it was returning nill, and even tried calling EXPLAIN inside another transaction but it didn't work either. Maybe I'm doing something wrong. Either way, we can leave it for last as an optimization task :)